### PR TITLE
graph: add algorithm for finding intervals in flow graphs

### PR DESCRIPTION
**Pull request by harpend**
Hi,

This is the code for issue #796 allowing finding intervals in flow graphs. I've added some basic testing using an example graph given in C. Cifuentes, "Reverse Compilation Techniques", 1994. This is my first time contributing to an open source project and I look forward to your feedback!

Thanks,
Jack

### CODE REVIEW

File: graph/flow/interval.go
Comment (kortschak) [inline graph/flow/interval.go:214]: This presumably wants a method to make use of the unexported `from` field?

**Review by kortschak — COMMENTED** [review]
Suggestions for code changes.
```diff
diff --git a/graph/flow/interval.go b/graph/flow/interval.go
index 1abcb562..0b0c9d0a 100644
--- a/graph/flow/interval.go
+++ b/graph/flow/interval.go
@@ -6,6 +6,7 @@ package flow
 
 import (
        "maps"
+       "slices"
 
        "gonum.org/v1/gonum/graph"
        "gonum.org/v1/gonum/graph/internal/linear"
@@ -41,29 +42,26 @@ terminates
 // Returns the interval graph given by the directed graph.
 // The interval graph contains the individual intervals.
 func Intervals(g graph.Directed, eid int64) IntervalGraph {
-       var worklist linear.NodeQueue
-       var intervals []*Interval
+       var intervals []Interval
+
        var ns linear.NodeStack
        visited := make(map[int64]bool)
-       inInterval := make(map[int64]graph.Node)
-
        dfsPostorder(g, eid, &ns, visited)
-       reversePostorderNodes := reversePostorder(ns)
-       n := g.Node(eid)
-       worklist.Enqueue(n)
-
-       for worklist.Len() > 0 {
+       slices.Reverse(ns)
+       var worklist linear.NodeQueue
+       worklist.Enqueue(g.Node(eid))
+       inInterval := make(map[int64]graph.Node)
+       for worklist.Len() != 0 {
                var interval Interval
-               n = worklist.Dequeue()
-               maps.Copy(inInterval, interval.findInterval(&n, g))
-               intervals = append(intervals, &interval)
+               maps.Copy(inInterval, interval.findInterval(worklist.Dequeue(), g))
+               intervals = append(intervals, interval)
 
-               for _, node := range reversePostorderNodes {
-                       if inInterval[(*node).ID()] != nil {
+               for _, node := range ns {
+                       if inInterval[node.ID()] != nil {
                                continue
                        }
 
-                       preds := g.To((*node).ID())
+                       preds := g.To(node.ID())
                        predsLength := preds.Len()
                        x := 0
                        for preds.Next() {
@@ -72,15 +70,20 @@ func Intervals(g graph.Directed, eid int64) IntervalGraph {
                                }
                        }
 
-                       if x > 0 && x < predsLength {
-                               worklist.Enqueue(*node)
+                       if 0 < x && x < predsLength {
+                               worklist.Enqueue(node)
                                break
                        }
                }
        }
 
-       ig := linkIntervals(intervals, g)
-       return ig
+       return linkIntervals(intervals, g)
+}
+
+// Contains the intervals and the edges between the intervals.
+type IntervalGraph struct {
+       Intervals []Interval
+       from      map[int64]map[int64]graph.Edge
 }
 
 // An Interval I(h) is the maximal, single entry subgraph for which h (head)
@@ -144,12 +147,12 @@ func (i *Interval) Node(id int64) graph.Node {
 // Finds all interval nodes.
 // Nodes are added to the interval if all their predecessors are in
 // the interval or they are the header node.
-func (i *Interval) findInterval(head *graph.Node, g graph.Directed) map[int64]graph.Node {
-       i.head = *head
+func (i *Interval) findInterval(h graph.Node, g graph.Directed) map[int64]graph.Node {
+       i.head = h
        var nq linear.NodeQueue
-       nq.Enqueue(*head)
+       nq.Enqueue(h)
        i.nodes = make(map[int64]graph.Node)
-       i.nodes[(*head).ID()] = *head
+       i.nodes[h.ID()] = h
        var node graph.Node
        for nq.Len() > 0 {
                node = nq.Dequeue()
@@ -193,31 +196,11 @@ func dfsPostorder(g graph.Directed, eid int64, ns *linear.NodeStack, visited map
                        dfsPostorder(g, succ.ID(), ns, visited)
                }
        }
-
-       n := g.Node(eid)
-       ns.Push(n)
-}
-
-// Extracts all nodes in the stack into an array in reverse postorder.
-func reversePostorder(ns linear.NodeStack) []*graph.Node {
-       var nodes []*graph.Node
-       stackLength := ns.Len()
-       for i := 0; i < stackLength; i++ {
-               n := ns.Pop()
-               nodes = append(nodes, &n)
-       }
-
-       return nodes
-}
-
-// Contains the intervals and the edges between the intervals.
-type IntervalGraph struct {
-       Intervals []*Interval
-       from      map[int64]map[int64]graph.Edge
+       ns.Push(g.Node(eid))
 }
 
 // Computes the internal and external edges for the intervals.
-func linkIntervals(intervals []*Interval, g graph.Directed) IntervalGraph {
+func linkIntervals(intervals []Interval, g graph.Directed) IntervalGraph {
        ig := IntervalGraph{
                Intervals: intervals,
                from:      make(map[int64]map[int64]graph.Edge),
diff --git a/graph/flow/interval_test.go b/graph/flow/interval_test.go
index 45ab91fd..56203734 100644
--- a/graph/flow/interval_test.go
+++ b/graph/flow/interval_test.go
@@ -9,57 +9,31 @@ import (
 func TestInterval(t *testing.T) {
        // Graph from C. Cifuentes, "Reverse Compilation Techniques", 1994 (figure 6-23)
        g := simple.NewDirectedGraph()
-       n1 := simple.Node(1)
-       n2 := simple.Node(2)
-       n3 := simple.Node(3)
-       n4 := simple.Node(4)
-       n5 := simple.Node(5)
-       n6 := simple.Node(6)
-       n7 := simple.Node(7)
-       n8 := simple.Node(8)
-       n9 := simple.Node(9)
-       n10 := simple.Node(10)
-       n11 := simple.Node(11)
-       n12 := simple.Node(12)
-       n13 := simple.Node(13)
-       n14 := simple.Node(14)
-       n15 := simple.Node(15)
-       g.AddNode(n1)
-       g.AddNode(n2)
-       g.AddNode(n3)
-       g.AddNode(n4)
-       g.AddNode(n5)
-       g.AddNode(n6)
-       g.AddNode(n7)
-       g.AddNode(n8)
-       g.AddNode(n9)
-       g.AddNode(n10)
-       g.AddNode(n11)
-       g.AddNode(n12)
-       g.AddNode(n13)
-       g.AddNode(n14)
-       g.AddNode(n15)
-
-       g.SetEdge(g.NewEdge(n1, n2))
-       g.SetEdge(g.NewEdge(n1, n5))
-       g.SetEdge(g.NewEdge(n2, n3))
-       g.SetEdge(g.NewEdge(n2, n4))
-       g.SetEdge(g.NewEdge(n3, n5))
-       g.SetEdge(g.NewEdge(n4, n5))
-       g.SetEdge(g.NewEdge(n5, n6))
-       g.SetEdge(g.NewEdge(n6, n7))
-       g.SetEdge(g.NewEdge(n6, n12))
-       g.SetEdge(g.NewEdge(n7, n8))
-       g.SetEdge(g.NewEdge(n7, n9))
-       g.SetEdge(g.NewEdge(n8, n9))
-       g.SetEdge(g.NewEdge(n8, n10))
-       g.SetEdge(g.NewEdge(n9, n10))
-       g.SetEdge(g.NewEdge(n10, n11))
-       g.SetEdge(g.NewEdge(n12, n13))
-       g.SetEdge(g.NewEdge(n14, n13))
-       g.SetEdge(g.NewEdge(n13, n14))
-       g.SetEdge(g.NewEdge(n14, n15))
-       g.SetEdge(g.NewEdge(n15, n6))
+       edges := []struct{ from, to int }{
+               {1, 2},
+               {1, 5},
+               {2, 3},
+               {2, 4},
+               {3, 5},
+               {4, 5},
+               {5, 6},
+               {6, 7},
+               {6, 12},
+               {7, 8},
+               {7, 9},
+               {8, 9},
+               {8, 10},
+               {9, 10},
+               {10, 11},
+               {12, 13},
+               {14, 13},
+               {13, 14},
+               {14, 15},
+               {15, 6},
+       }
+       for _, e := range edges {
+               g.SetEdge(g.NewEdge(simple.Node(e.from), simple.Node(e.to)))
+       }
 
        // test number of intervals
        ig := Intervals(g, 1)
```

**Comment by harpend** [comment]
Added the changes suggested

Also switched IntervalGraph to use to instead of from so the nodes to an interval header node can be accessed.
This is accessed through the To function.

File: graph/flow/interval_test.go
Comment (kortschak) [inline graph/flow/interval_test.go:1]: This needs a copyright header.

File: graph/flow/interval_test.go
Comment (mewmew) [inline graph/flow/interval_test.go:49]: Consider extending the test cases to also verify that the correct nodes are added to each interval.

**Review by mewmew — CHANGES_REQUESTED** [review]
LGTM, after minor extension of the test cases to also verify that the correct nodes are added to each interval.

Spectacular work @harpend!!

File: graph/flow/interval_test.go
Comment (kortschak) [inline graph/flow/interval_test.go:51]: Why is this being done?

Comment (harpend) [reply graph/flow/interval_test.go]: Forgot to remove it, all good now.

File: graph/flow/interval_test.go
Comment (mewmew) [inline graph/flow/interval_test.go:74]: should the error message read "not found"?

Comment (harpend) [reply graph/flow/interval_test.go:74]: yes you are correct

**Comment by kortschak** [comment]
I'd like to go over the tests and make them more table driven, but I have not yet had time.

**Comment by kortschak** [comment]
I've gone through the tests to make them accept new test vectors when the need arises, and to more fully test the vectors. This raised issues for me with the semantics here.

The `Interval` type implements `graph.Graph`, but it looks to me like it is formally a directed graph. It would be nice to include this in the type if this is not too much work/weight.

In order to get the more rigorous testing to work and the `Interval` type to work with the rest of the graph packages, I needed to make the `IntervalGraph` hold and return `*Interval`. I'm not particularly fussed by this.

```diff
diff --git a/graph/flow/interval.go b/graph/flow/interval.go
index 5dcda8c7..73ceb1ed 100644
--- a/graph/flow/interval.go
+++ b/graph/flow/interval.go
@@ -48,7 +48,7 @@ terminates
 // of the directed graph in g from starting from the entry node identified
 // by eid.
 func Intervals(g graph.Directed, eid int64) IntervalGraph {
-       var intervals []Interval
+       var intervals []*Interval
 
        var ns linear.NodeStack
        visited := make(map[int64]bool)
@@ -62,7 +62,7 @@ func Intervals(g graph.Directed, eid int64) IntervalGraph {
        for worklist.Len() != 0 {
                var interval Interval
                maps.Copy(inInterval, interval.findInterval(worklist.Dequeue(), g))
-               intervals = append(intervals, interval)
+               intervals = append(intervals, &interval)
 
                for _, node := range ns {
                        if inInterval[node.ID()] != nil {
@@ -90,7 +90,7 @@ func Intervals(g graph.Directed, eid int64) IntervalGraph {
 
 // IntervalGraph contains the intervals and the edges between the intervals.
 type IntervalGraph struct {
-       Intervals []Interval
+       Intervals []*Interval
        headers   map[int64]graph.Node
        to        map[int64]map[int64]graph.Edge
 }
@@ -110,7 +110,7 @@ func (ig *IntervalGraph) To(id int64) graph.Nodes {
 type Interval struct {
        head  graph.Node
        nodes map[int64]graph.Node
-       from  map[int64]map[int64]graph.Edge
+       edges map[int64]map[int64]graph.Edge
 }
 
 // Head returns header node for an interval.
@@ -129,7 +129,7 @@ func (i *Interval) Nodes() graph.Nodes {
 // Edge returns the edge given 2 node id's if the edge exists.
 // Else it returns null.
 func (i *Interval) Edge(uid, vid int64) graph.Edge {
-       edge, ok := i.from[uid][vid]
+       edge, ok := i.edges[uid][vid]
        if !ok {
                return nil
        }
@@ -140,19 +140,19 @@ func (i *Interval) Edge(uid, vid int64) graph.Edge {
 // The returned graph.Nodes is only valid until the next mutation of
 // the receiver.
 func (i *Interval) From(id int64) graph.Nodes {
-       if len(i.from[id]) == 0 {
+       if len(i.edges[id]) == 0 {
                return graph.Empty
        }
-       return iterator.NewNodesByEdge(i.nodes, i.from[id])
+       return iterator.NewNodesByEdge(i.nodes, i.edges[id])
 }
 
 // HasEdgeBetween returns whether an edge exists between nodes x and y without
 // considering direction.
 func (i *Interval) HasEdgeBetween(xid int64, yid int64) bool {
-       if _, ok := i.from[xid][yid]; ok {
+       if _, ok := i.edges[xid][yid]; ok {
                return true
        }
-       _, ok := i.from[yid][xid]
+       _, ok := i.edges[yid][xid]
        return ok
 }
 
@@ -219,7 +219,7 @@ func dfsPostorder(g graph.Directed, eid int64, ns *linear.NodeStack, visited map
 }
 
 // linkIntervals computes the internal and external edges for the intervals.
-func linkIntervals(intervals []Interval, g graph.Directed) IntervalGraph {
+func linkIntervals(intervals []*Interval, g graph.Directed) IntervalGraph {
        ig := IntervalGraph{
                Intervals: intervals,
                headers:   make(map[int64]graph.Node),
@@ -228,17 +228,17 @@ func linkIntervals(intervals []Interval, g graph.Directed) IntervalGraph {
 
        for i, interval := range intervals {
                ig.headers[int64(i)] = interval.head
-               interval.from = make(map[int64]map[int64]graph.Edge)
+               interval.edges = make(map[int64]map[int64]graph.Edge)
                for _, node := range interval.nodes {
                        succs := g.From(node.ID())
                        for succs.Next() {
                                succNode := succs.Node()
                                if interval.nodes[succNode.ID()] != nil {
-                                       if interval.from[node.ID()] == nil {
-                                               interval.from[node.ID()] = make(map[int64]graph.Edge)
+                                       if interval.edges[node.ID()] == nil {
+                                               interval.edges[node.ID()] = make(map[int64]graph.Edge)
                                        }
 
-                                       interval.from[node.ID()][succNode.ID()] = g.Edge(node.ID(), succNode.ID())
+                                       interval.edges[node.ID()][succNode.ID()] = g.Edge(node.ID(), succNode.ID())
                                } else {
                                        if ig.to[succNode.ID()] == nil {
                                                ig.to[succNode.ID()] = make(map[int64]graph.Edge)
diff --git a/graph/flow/interval_test.go b/graph/flow/interval_test.go
index acfc96c5..692ec2b7 100644
--- a/graph/flow/interval_test.go
+++ b/graph/flow/interval_test.go
@@ -7,77 +7,127 @@ package flow
 import (
        "testing"
 
+       "gonum.org/v1/gonum/graph"
+       "gonum.org/v1/gonum/graph/encoding/dot"
        "gonum.org/v1/gonum/graph/simple"
+       "gonum.org/v1/gonum/graph/topo"
 )
 
-func TestInterval(t *testing.T) {
-       // Graph from C. Cifuentes, "Reverse Compilation Techniques", 1994 (figure 6-23)
-       g := simple.NewDirectedGraph()
-       edges := []struct{ from, to int }{
-               {1, 2},
-               {1, 5},
-               {2, 3},
-               {2, 4},
-               {3, 5},
-               {4, 5},
-               {5, 6},
-               {6, 7},
-               {6, 12},
-               {7, 8},
-               {7, 9},
-               {8, 9},
-               {8, 10},
-               {9, 10},
-               {10, 11},
-               {12, 13},
-               {14, 13},
-               {13, 14},
-               {14, 15},
-               {15, 6},
-       }
-
-       nodes := [][]int64{
-               {1, 2, 3, 4, 5},
-               {6, 7, 8, 9, 10, 11, 12},
-               {13, 14, 15},
-       }
+var intervalTests = []struct {
+       name  string
+       edges []struct{ from, to simple.Node }
+       root  int64
+       want  []map[int64]intset
+}{
+       {
+               // Graph from C. Cifuentes, "Reverse Compilation Techniques", 1994 (figure 6-23)
+               // Available via https://eprints.qut.edu.au/36820/
+               name: "cifuentes",
+               edges: []struct{ from, to simple.Node }{
+                       {1, 2},
+                       {1, 5},
+                       {2, 3},
+                       {2, 4},
+                       {3, 5},
+                       {4, 5},
+                       {5, 6},
+                       {6, 7},
+                       {6, 12},
+                       {7, 8},
+                       {7, 9},
+                       {8, 9},
+                       {8, 10},
+                       {9, 10},
+                       {10, 11},
+                       {12, 13},
+                       {14, 13},
+                       {13, 14},
+                       {14, 15},
+                       {15, 6},
+               },
+               root: 1,
+               want: []map[int64]intset{
+                       {
+                               1: linksTo(2, 5),
+                               2: linksTo(3, 4),
+                               3: linksTo(5),
+                               4: linksTo(5),
+                               5: nil,
+                       },
+                       {
+                               6:  linksTo(7, 12),
+                               7:  linksTo(8, 9),
+                               8:  linksTo(9, 10),
+                               9:  linksTo(10),
+                               10: linksTo(11),
+                               11: nil,
+                               12: nil,
+                       },
+                       {
+                               13: linksTo(14),
+                               14: linksTo(13, 15),
+                               15: nil,
+                       },
+               },
+       },
+}
 
-       intervalSize := []int64{
-               5, 7, 3,
-       }
+func TestInterval(t *testing.T) {
+       for _, test := range intervalTests {
+               t.Run(test.name, func(t *testing.T) {
+                       g := simple.NewDirectedGraph()
+                       for _, e := range test.edges {
+                               g.SetEdge(simple.Edge{F: e.from, T: e.to})
+                       }
 
-       for _, e := range edges {
-               g.SetEdge(g.NewEdge(simple.Node(e.from), simple.Node(e.to)))
-       }
+                       ig := Intervals(g, test.root)
+                       if len(ig.Intervals) != len(test.want) {
+                               t.Fatalf("unexpected interval count: got:%d want:%d", len(ig.Intervals), len(test.want))
+                       }
 
-       // Test number of intervals
-       ig := Intervals(g, 1)
-       if len(ig.Intervals) != 3 {
-               t.Fatalf("Expected 3 intervals, got %d", len(ig.Intervals))
+                       for i, iv := range ig.Intervals {
+                               var got graph.Graph = iv
+                               want := gFromIntsets(test.want[i])
+                               if !topo.Equal(got, want) {
+                                       gotDot, err := dot.Marshal(iv, "", "", "\t")
+                                       if err != nil {
+                                               t.Fatalf("unexpected error marshalling got DOT: %v", err)
+                                       }
+                                       wantDot, err := dot.Marshal(want, "", "", "\t")
+                                       if err != nil {
+                                               t.Fatalf("unexpected error marshalling want DOT: %v", err)
+                                       }
+                                       t.Errorf("unexpected topology of interval %d:\ngot:\n%s\nwant:\n%s", i, gotDot, wantDot)
+                               }
+                       }
+               })
        }
+}
 
-       for i, interval := range ig.Intervals {
-               // Test correct number of nodes are found in interval
-               if len(interval.nodes) != int(intervalSize[i]) {
-                       t.Errorf("Expected %d nodes in interval %d, got %d", intervalSize[i], i, len(interval.nodes))
+func gFromIntsets(s map[int64]intset) graph.Directed {
+       g := simple.NewDirectedGraph()
+       for u, e := range s {
+               // Add nodes that are not defined by an edge.
+               if g.Node(int64(u)) == nil {
+                       g.AddNode(simple.Node(u))
                }
-
-               // Test all expected in interval are present
-               for _, node := range nodes[i] {
-                       if interval.nodes[node] == nil {
-                               t.Errorf("Node %d not found in interval %d", node, i)
-                       }
+               for v := range e {
+                       g.SetEdge(simple.Edge{F: simple.Node(u), T: simple.Node(v)})
                }
        }
+       return g
+}
 
-       // test interval edges
-       toEdges := ig.To(6)
-       if toEdges.Len() != 2 {
-               t.Errorf("Expected 2 edges to node 6 in interval graph, got %d", len(ig.to))
-       }
+// intset is an integer set.
+type intset map[int64]struct{}
 
-       toEdges2 := ig.To(13)
-       if toEdges2.Len() != 1 {
-               t.Errorf("Expected 1 edge to node 13 in interval graph, got %d", len(ig.to))
+func linksTo(i ...int64) intset {
+       if len(i) == 0 {
+               return nil
+       }
+       s := make(intset)
+       for _, v := range i {
+               s[v] = struct{}{}
        }
+       return s
 }
```

**Comment by harpend** [comment]
Interval is updated to be a directed graph, is there any desire for interval graph to be a directed graph as well?

**Comment by kortschak** [comment]
I think that would make sense.

**Comment by harpend** [comment]
added tests for edges between intervals and IntervalGraph now implements directed graph

**Review by mewmew — APPROVED** [review]

**Comment by mewmew** [comment]
With the updated tests and the change to have interval implement the directed graph interface, I think the PR looks good.

@kortschak has the last say before merging : )

Really happy to see this finally being added to Gonum. Awesome work @harpend!

Cheerful regards,
Robin

**Review by kortschak — COMMENTED** [review]
nit only then LGTM

We will also need another, separate, PR with the commit message "A+C: add \<YOUR NAME\>" adding yourself to AUTHORS and CONTRIBUTORS.

**Comment by harpend** [comment]
> I'll take a look on the weekend. Are you able to send the A+C before then? If so, I can merge then as well.

A+C PR is already opened.

**Comment by harpend** [comment]
The email in the header I only use for github, and do not wish to include a personal email.

**Comment by kortschak** [comment]
OK

**Review by kortschak — APPROVED** [review]
Thanks

**Pull request merged**
